tests/posix/common/src/key.c

```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
 	}
 }
 
+static void *setspecific_thread(void *count)
+{
+	int value = 42;
+	int *alloc_count = count;
+
+	while (1) {
+		pthread_key_t key;
+
+		zassert_ok(pthread_key_create(&key, NULL), "failed to create key");
+		if (pthread_setspecific(key, &value) == ENOMEM) {
+			break;
+		};
+		*alloc_count += 1;
+	}
+
+	return NULL;
+}
+
+ZTEST(key, test_thread_specific_data_deallocation)
+{
+	pthread_t thread;
+	int alloc_count_t0 = 0;
+	int alloc_count_t1 = 0;
+
+	zassert_ok(pthread_create(&thread, NULL, setspecific_thread, &alloc_count_t0),
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
```

Comment:
Maybe before spawning the second thread, add a `printk()` that displays how many times a new key was allocated and assigned a value.